transport/transport.ts

```diff
@@ -515,3 +476,220 @@ export abstract class Transport<ConnType extends Connection> {
+      }
+
+      // send boot sequence
+      this.state = 'open';
+      const responseMsg = msg(this.clientId, to, nanoid(), {
+        type: 'HANDSHAKE_REQ',
+        protocolVersion: PROTOCOL_VERSION,
+      } satisfies Static<typeof ControlMessageHandshakeRequestSchema>);
+      conn.send(this.codec.toBuffer(responseMsg));
+    } catch (err: unknown) {
+      // retry on failure
+      this.inflightConnectionPromises.delete(to);
+      if (attempt >= this.options.retryAttemptsMax) {
+        throw new Error(
+          `${this.clientId} -- connection to ${to} failed after ${attempt} attempts (${err}), giving up`,
+        );
+      } else {
+        // exponential backoff + jitter
+        const jitter = Math.floor(Math.random() * this.options.retryJitterMs);
+        const backoffMs = this.options.retryIntervalMs * 2 ** attempt + jitter;
+        log?.warn(
+          `${this.clientId} -- connection to ${to} failed (${err}), trying again in ${backoffMs}ms`,
+        );
+        setTimeout(() => this.connect(to, attempt + 1), backoffMs);
+      }
+    }
+  }
+
+  receiveWithBootSequence(
+    conn: ConnType,
+    sessionCb?: (sess: Session<ConnType>) => void,
+  ) {
+    const bootHandler = (data: Uint8Array) => {
+      const parsed = this.parseMsg(data);
+      if (!parsed) return;
+
+      if (!Value.Check(ControlMessageHandshakeResponseSchema, parsed.payload)) {
+        log?.warn(
+          `${
+            this.clientId
+          } -- received invalid handshake resp: ${JSON.stringify(parsed)}`,
+        );
+        return;
+      }
+
+      if (!parsed.payload.status.ok) {
+        log?.warn(
+          `${this.clientId} -- received failed handshake resp: ${JSON.stringify(
+            parsed,
+          )}`,
+        );
+        return;
+      }
+
+      // everything is ok
+      // connect the session
+      const session = this.onConnect(conn, parsed.from);
+      sessionCb?.(session);
+      conn.removeDataListener(bootHandler);
+      conn.addDataListener((data) => this.handleMsg(this.parseMsg(data)));
+    };
+
+    return bootHandler;
+  }
+
+  onDisconnect(conn: ConnType, connectedTo: string | undefined): void {
+    if (connectedTo) this.inflightConnectionPromises.delete(connectedTo);
+    super.onDisconnect(conn, connectedTo);
+  }
+}
+
+export abstract class ServerTransport<
+  ConnType extends Connection,
+> extends Transport<ConnType> {
+  protected handleConnection(conn: ConnType) {
+    let session: Session<ConnType> | undefined = undefined;
+    const client = () => session?.connectedTo ?? 'unknown';
+    conn.addDataListener(
+      this.receiveWithBootSequence(conn, (establishedSession) => {
+        session = establishedSession;
+      }),
+    );
+
+    conn.addCloseListener(() => {
+      if (!session) return;
+      log?.info(
+        `${this.clientId} -- connection (id: ${
+          conn.debugId
+        }) to ${client()} disconnected`,
+      );
+      this.onDisconnect(conn, session?.connectedTo);
```

Comment:
```suggestion
      this.onDisconnect(conn, session.connectedTo);
```

already checked if session is truthy